src/mlpack/core/cereal/pointer_wrapper.hpp

```diff
@@ -0,0 +1,64 @@
+#ifndef CEREAL_PONTER_WRAPPER_HPP
+#define CEREAL_PONTER_WRAPPER_HPP
+
+/*
+ * The objective of this class is to create a wrapper for
+ * raw pointer by encapsulating them in a smart pointer unique_ptr
+ * This will allow to serialize raw pointer in cereal as a smart pointer
+ * because it will be difficult to change all pointer type in mlpack
+ */
+
+#include <cereal/archives/binary.hpp>
+#include <cereal/archives/json.hpp>
+#include <cereal/archives/portable_binary.hpp>
+#include <cereal/archives/xml.hpp>
+#include <cereal/types/memory.hpp>
+
+namespace cereal {
+
+template<class T>
+class pointer_wrapper
+{
+public:
+
+   pointer_wrapper(T * pointer)
+    {  
+        localPointer = std::move(pointer);
+        std::cout << "address of local ptr" << localPointer << std::endl;
+        std::cout << "address of  pointer" << pointer << std::endl;
+    }
+    
+  template<class Archive>
+  void save(Archive& ar, const unsigned int /*version*/) const
+  {
+    std::unique_ptr<T> smartPointer = std::make_unique<T>(*this->localPointer);
+    ar(CEREAL_NVP(smartPointer));
+  }
+
+  template<class Archive>
+  void load(Archive& ar, const unsigned int /*version*/)
+  {
+      std::unique_ptr<T> smartPointer;
+      ar(CEREAL_NVP(smartPointer));
+      this->localPointer = smartPointer.release();
+      std::cout << "The value of localPointer is : " << *localPointer << std::endl;
+      std::cout << "The addresss of localPointer is : " << &localPointer << std::endl;
```

Comment:
@robertohueso @rcurtin The save function is working fine, as the loading function. However, when we loading the address is recovered by the local pointer data member. I would like to be a able to update also the address pointed by localPointer into the pointer passed by the constructor.

Reply:
Maybe `localPointer` needs to be a double-pointer I am not sure, I will give it a try

Reply:
> Maybe `localPointer` needs to be a double-pointer I am not sure, I will give it a try

Exactly it can either be a double pointer
```c++
pointer_wrapper(T** pointer)
```
or pass a pointer by reference which I think might be safer :thinking: 
```c++
pointer_wrapper(T*& pointer)
```
It's probably also a good idea to check if it points to something so we can avoid memory leaks e.g.
```c++
template<class Archive>
void load(Archive& ar, const unsigned int /*version*/) {
  ...
  if (localPointer != nullptr)
    delete localPointer;
  ...
}
```

Reply:
@robertohueso when passing by reference `T*&` or `**` is giving a segfault.
If `pointer` was a `nullptr` is will segfault also.
The reason behind this is that when `load` is finishing, the address pointed by `unique_ptr` is being destroyed.

The only working solution is to initialize `pointer` before being passed to the wrapper and to change the value of what is pointing too by dereferencing directly the `localPointer` like this

Suppose `localPointer` is defined as this `T*& localpointer`
```
*this->localPointer = *smartPointer;
```
This will work, However, we are not sure that all pointer in mlpack are already initialized to something other than `nullptr`

Reply:
Just trying the print the value of the addresses I discovered that I am getting different addresses for `localPointer` when it is printed from inside `load` and from inside the constructor

Reply:
There seems to be something strange going on here---even if I comment out the `ar(CEREAL_NVP(smart_pointer))` from `pointer_wrapper::save()` (and actually comment out all of the `unique_ptr` code entirely, so that no `unique_ptr` is ever made), the program still segfaults if I allocate a new value for `pointer`:

```
  template<class Archive>
  void load(Archive& ar, const unsigned int /*version*/)
  {   
      this->localPointer = new T();
      std::cout << "The value of localPointer is : " << *localPointer << std::endl;   
      std::cout << "The addresss of localPointer is : " << &localPointer << std::endl;
  }
```

This is with `localPointer` of type `T*&`.

I don't think that the address being pointed to by `localPointer` is being destroyed by the `unique_ptr`; I think it's something else.  You could confirm or refute that idea by creating a `pointer_wrapper` around a class that has a destructor, and simply printing from the destructor, to see if it's called twice for a particular memory location. :+1:

Reply:
Interesting :thinking:, I was not able to reproduce this error for the same `localPointer`

Reply:
Hmm, I'm not sure why you weren't able to reproduce it.  In any case I'll resolve this and we can use the other comment thread. :+1:

Reply:
I tried playing with it a little bit more and the double pointer approach seems to work:

```
template<class T>
class pointer_wrapper
{
public:
  pointer_wrapper(T** pointer)
    : localPointer(pointer)
  {}

  template<class Archive>
  void save(Archive& ar, const unsigned int /*version*/) const
  {
    std::unique_ptr<T> smartPointer = std::make_unique<T>(*this->localPointer);
    ar(CEREAL_NVP(smartPointer));
  }

  template<class Archive>
  void load(Archive& ar, const unsigned int /*version*/)
  { 
    std::unique_ptr<T> smartPointer;
    ar(CEREAL_NVP(smartPointer));
    *localPointer = smartPointer.release();
  }

  T* release() { return *localPointer; }

private:
  T** localPointer;
};
```

This appears to have no issues.  Can you try that and see what it does for you?  Maybe the `T*&` was not being handled correctly, and there is something about the semantics of a pointer reference that mean that the original code wasn't right.